substrate/frame/examples/storage-migrations/src/lib.rs

```diff
@@ -0,0 +1,284 @@
+// This file is part of Substrate.
+
+// Copyright (C) Parity Technologies (UK) Ltd.
+// SPDX-License-Identifier: Apache-2.0
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! # Storage Migrations Example Pallet
+//!
+//! An example pallet explaining why storage migrations are necessary and demonstrating
+//! best-practices for writing them.
+//!
+//! It is intended to be a resource for guiding a developer from knowing nothing about storage
+//! migrations to being able to write safe, well tested, versioned migrations for their own pallets.
+//!
+//! ## Prerequisites
+//!
+//! Before writing pallet storage migrations, you should be familiar with:
+//! - [`Runtime upgrades`](https://docs.substrate.io/maintain/runtime-upgrades/)
+//! - The [`GetStorageVersion`](frame_support::traits::GetStorageVersion) trait, and the difference
+//!   between current and on-chain [`StorageVersion`]s
+
+//! ## How to read these docs
+//! - Run `cargo doc --features try-runtime --package pallet-example-storage-migrations --open`
+//! to view the documentation in your browser.
+//! - Read the relevant source code as your read the docs.
+//!
+//! ## Pallet Overview
+//!
+//! This example pallet contains a single storage item [`Value`](pallet::Value), which may be set by
+//! any signed origin by calling the [`set_value`](crate::Call::set_value) extrinsic.
+//!
+//! For the purposes of this exercise, we imagine that in [`StorageVersion`] V0 of this pallet
+//! [`Value`](pallet::Value) is a `u32`, and this what is currently stored on-chain.
+//!
+//! ```ignore
+//! // V0 Storage Value
+//! pub type Value<T: Config> = StorageValue<_, u32>;
+//! ```
+//!
+//!
+//! In [`StorageVersion`] V1 of the pallet a new struct [`CurrentAndPreviousValue`] is introduced:
+//!
+//! ```ignore
+//! pub struct CurrentAndPreviousValue {
+//! 	/// The most recently set value.
+//! 	pub current: u32,
+//! 	/// The previous value, if one existed.
+//! 	pub previous: Option<u32>,
+//! }
+//! ```
+//!
+//! and [`Value`](pallet::Value) is updated to store this new struct instead of a `u32`:
+//!
+//! ```ignore
+//! // V1 Storage Value
+//! pub type Value<T: Config> = StorageValue<_, CurrentAndPreviousValue>;
```

Comment:
and this etc.

Reply:
docify embeds were not compiling in this file until Sam pushed a fix yesterday. already have a TODO mentioned in the PR description to use it once it's fixed :)